### PR TITLE
Basic VM example

### PR DESCRIPTION
**Pull request by abdulbahajaj**
Hey, here is a UVM Basic VM. It is a REPL for a simple Basic dialect with a canvas that you can use to draw stuff.

image

### CODE REVIEW

File: ncc/examples/basic.c [hunk 1]
```diff
@@ -0,0 +1,1929 @@
...
+	for (int i = 0; i < NUM_COLS; ++i)
+	{
+		if (text[row_idx][i] == 0)
+		{
+			return i;
+		}
+	}
+
+	return NUM_COLS;
+}
```
Comment (maximecb) [inline ncc/examples/basic.c:70]: Could I ask you to convert this to 4 spaces per indentation, no tabs?

File: ncc/examples/basic.c [hunk 2]
```diff
@@ -0,0 +1,1929 @@
...
+			if(idx >= vm_intern_buffer) idx = 0;
+		}
+		DEBUG("Found a spot at index: ");
+		DEBUGI(idx);
+		DEBUG("");
+
+		vm_intern_buffer[idx] = sym_meta;
+	}
+
+
```
Comment (maximecb) [inline ncc/examples/basic.c:1150]:
```suggestion
```

File: ncc/examples/basic.c [hunk 3]
```diff
@@ -0,0 +1,1929 @@
...
+			break;
+		}
+	}
+
+	DEBUG("Not enough space in intern buffer");
+	resize_interned_sym_buffer();
+	DEBUG("Done resizing intern buffer");
+	return vm_intern(sym, len, hash);
+
+
```
Comment (maximecb) [inline ncc/examples/basic.c:1203]:
```suggestion


```

File: ncc/examples/basic.c (hunk 1 above)
Comment (abdulbahajaj) [reply ncc/examples/basic.c:70]: yes on it.

Comment (abdulbahajaj) [reply ncc/examples/basic.c]: done

File: ncc/examples/basic.c [hunk 4]
```diff
@@ -0,0 +1,1927 @@
...
+}
+
+u64 vm_alloc_var()
+{
+    u64 allocated = vm_vars_allocated;
+    ++vm_vars_allocated;
+    return allocated;
+}
+
+int strcmp_with_len(char* a, char* b, size_t b_len)
```
Comment (maximecb) [inline ncc/examples/basic.c:1070]: I believe this is equivalent to `strncmp`: https://cplusplus.com/reference/cstring/strncmp/

We should probably implement this function in `include/string.h` and use that.

File: ncc/examples/basic.c [hunk 5]
```diff
@@ -0,0 +1,1927 @@
...
+            return -1;
+        else if (ch_a > ch_b)
+            return 1;
+    }
+    return 0;
+}
+
+#define SYM_META_STR 0
+#define SYM_META_LOC 1
+#define SYM_META_LOOP_PATCH 2
```
Comment (maximecb) [inline ncc/examples/basic.c:1091]: Should comment to explain what these are for :)

**Comment by maximecb** [comment]
I just tried your program. Sorry I didn't get back to you faster! I think your basic program is pretty great. Going to give you small feedback from a usability standpoint:

1. This is a nitpick/stylistic: the console says "type help for available c
ommands". I wish that commands was on its own line instead of having the c on a separate line.

2. I wish that there was an "EXIT" command. Imo people are likely to reach for that. Would make it more intuitive. May want to also support "QUIT".

3. I wish that the prompt started with a `>` or `]` or `$` character. Would make it easier to distinguish the commands I type from the printed output.

4. Could be nice to have a FILL command to fill the entire canvas with a given color? :)

5. There is no error message when an unsupported command is entered. It just says 0, then ready? I was expecting PRINT "HI" to work, but it wasn't, and I wasn't sure why.

6. Supported color names in the help are in lowercase, but the color names you expect are actually uppercase RED, BLUE, etc.

7. I couldn't really figure out how to make an infinite loop happen with GOTO. Hmmm.

I'm being very nitpicky here. I think your basic program is already impressive, but there's definitely lots of little details of polish that could be added and together would make it even stronger and more intuitive to use.

**Comment by maximecb** [comment]
I added an `strncmp()` function to `include/string.h` so you can make use of it if useful :)

https://github.com/maximecb/uvm/commit/4153cf10494c87fbb8e135b0afe9143942e8fa80

**Comment by abdulbahajaj** [comment]
Appreciate the feedback
Will do/fix 1, 2, 3, 4, 6
For 5 when you enter an unrecognized command like say hello it will treat it as a variable and print its value.
so "LET a 1" will change the value of a to 1. Then typing a will print 1. So when you wrote PRINT, it thought that it was a variable, which had the value 0 since you did not initialize it.

For 7 you can have an infinite loop by loading the command
There are two ways of entering commands: a one off command that gets executed and never runs again and a "loaded" command. The latter can be done by prefixing the command with a number. After loading a command you can run using the RUN command.
For example

image

**Comment by maximecb** [comment]
> For 5 when you enter an unrecognized command like say hello it will treat it as a variable and print its value.
so "LET a 1" will change the value of a to 1. Then typing a will print 1. So when you wrote PRINT, it thought that it was a variable, which had the value 0 since you did not initialize it.

It's your basic VM so I respect your choices, but maybe looking up uninitialized variables should print an error instead of evaluating to zero? Would be more intuitive for new users? Or is that just the way basic typically works?

**Comment by abdulbahajaj** [comment]
Ya I can see how it could be confusing based on ur feedback. Will update it to throw an error message.

**Comment by abdulbahajaj** [comment]
I have completed everything we discussed. Let me know if I missed something.

Feel free to give more suggestions!

**Comment by maximecb** [comment]
One thing that seems like it would be an improvement would be to update the UI every N iterations.

At the moment if you do a loop like:
10 "hi"
20 GOTO 10

You don't see anything update because it's an infinite loop, but this isn't how "old school" basic would behave. Same if you did a loop to plot things.

Another potential improvement would be to add a PRINT command. My understanding is that your basic basically doesn't need that, but I think people will expect it to be there, even if it's not strictly necessary.

Another thing would be to maybe have some random number generator 🤔 

Those are 3 things I can think of. All of that being said, I'm happy to merge your current version if you want. It is already a very cool demo 👌 🙂

**Comment by abdulbahajaj** [comment]
oh cool I see you merged it 🎉 🎉 🎉 🎉 
I have been working on adding your ideas. They are all done. I have also added a sleep command that allow users to delay the time in between commands.
Currently I am trying to figure out a good UX for help since there are just too many commands now that they won't fit on the screen.
 I should be able to push something today. Will create a new PR since this one is merged.

**Comment by abdulbahajaj** [comment]
I have also added a way to escape infinite loops by pressing escape ( you don't need to do that for the cavas to update tho)

Here is a cool demo!

image

**Comment by maximecb** [comment]
Sweet!!! :D

**Pull request merged**